### PR TITLE
[SYSADMIN ACTION][Feature] Preferred Name Logging

### LINKED ISSUE #3273: Preferred name is forgotten/removed without any indication why
I just received a report from a current RPI user that their preferred name was removed. This user had set a preferred first name at the beginning of the semester and discovered this week that their name on Submitty was changed back to their legal name. The name did not violate the rules stated in the 'Specify Preferred Names' pop-up and is the name the user goes by in regular/day-to-day life and on paper/in-class submissions.

It's unclear if a professor changed it (though the pop-up does indicate that is possible), but I wouldn't expect that in this case. Is there a good way for a user to determine why their name was changed, or any recent actions taken that may have unintentionally cleared the field? 

(This particular user was also super excited to see this feature existed after noticing the pencil icon next to their name, it's awesome to see inclusive feature work like this!)

### PR DESCRIPTION
**Pull request by pbailie**
### NOTE: SYSADMIN ACTION REQUIRED

After this PR is installed, root access is required to manually restart Postgresql.  

`service postgresql restart`

---------------

Submitty code portion of logging changes to a user's preferred first and/or last names.

- Database user update query notes who was logged in as database cannot know (all queries are run by the same "role").
- Update database trigger function to detect and log when user preferred name(s) have changed.
- Created migration for new trigger function code

Also needed is a sysadmin tools script that will scrape the logs created by Postgresql and create its own log of all (and only) preferred name changes.  (q.v. Submitty/SysadminTools#16)

Close #3273

### CODE REVIEW

**Comment by MasterOdin** [comment]
Is the only place preferred name updated within just that function?

**Comment by pbailie** [comment]
> Is the only place preferred name updated within just that function?

So far as I can tell.  I grepped for `user_preferred_firstname` and `user_preferred_lastname` among the database queries code files.  Starting with `DatabaseQueries.php`:

```
grep -n 'user_preferred_firstname' DatabaseQueries.php 
2151:          user_preferred_firstname, user_lastname, late_day_exceptions
2442:      $this->course_db->query("SELECT user_firstname, user_preferred_firstname, user_lastname, user_preferred_lastname, user_email FROM users WHERE user_id = ?", array($user_id));
2445:      $ar["first_name"] = (empty($name_rows["user_preferred_firstname"])) ? $name_rows["user_firstname"]      : $name_rows["user_preferred_firstname"];
```
```
grep -n 'user_preferred_lastname' DatabaseQueries.php 
2442:      $this->course_db->query("SELECT user_firstname, user_preferred_firstname, user_lastname, user_preferred_lastname, user_email FROM users WHERE user_id = ?", array($user_id));
2446:      $ar["last_name"]  = (empty($name_rows["user_preferred_lastname"]))  ? " " . $name_rows["user_lastname"] : " " . $name_rows["user_preferred_lastname"];
```
- Line 2151 is part of `getUsersWithExtensions()`.
- Lines 2442, 2445, 2446 are part of `getDisplayUserInfoFromUserId()`.

Next, `PostgresqlDatabaseQueries.php`:
```
$ grep -n 'user_preferred_firstname' PostgresqlDatabaseQueries.php 
122:        $this->submitty_db->query("INSERT INTO users (user_id, user_password, user_numeric_id, user_firstname, user_preferred_firstname, user_lastname, user_preferred_lastname, user_email, user_updated, instructor_updated)
168:  user_numeric_id=?, user_firstname=?, user_preferred_firstname=?,
461:        SELECT u.user_id, user_firstname, user_preferred_firstname,
969:              u.user_preferred_firstname,
1040:              gcd.array_grader_user_preferred_firstname,
1104:                  json_agg(ug.user_preferred_firstname) AS array_grader_user_preferred_firstname,
1244:                'user_preferred_firstname',
```
```
$ grep -n 'user_preferred_lastname' PostgresqlDatabaseQueries.php 
122:        $this->submitty_db->query("INSERT INTO users (user_id, user_password, user_numeric_id, user_firstname, user_preferred_firstname, user_lastname, user_preferred_lastname, user_email, user_updated, instructor_updated)
169:  user_lastname=?, user_preferred_lastname=?,
462:          user_lastname, user_preferred_lastname, allowed_late_days, since_timestamp::timestamp::date
971:              u.user_preferred_lastname,
```
- Line 122 is part of `insertSubmittyUser()`.
- Lines 168, 169 are part of `updateUser()` — the update code I worked on.
- Lines 461, 462 are part of `getUsersWithLateDays()`.
- Lines 969, 971, 1040, 1104, 1244 are part of `getGradedGradeablesUserOrTeam()`.

`updateUser()` appears to be the only function an update/change is performed on preferred name data.

**Comment by MasterOdin** [comment]
Given that this only happens here and in the student feed, would it be easier to just have this write the lines into a log file directly without going through the database and relying on query comments (which I fear are too easy to lose on a refactor).

**Comment by pbailie** [comment]
> Given that this only happens here and in the student feed, would it be easier to just have this write the lines into a log file directly without going through the database and relying on query comments (which I fear are too easy to lose on a refactor).

I disagree.  To convert everything over to the PHP level runs greater risk of "easy to lose on a refactor" than having the logging done on the database level.

Having the database log the change will ensure that any PHP code refactor won't disable the logging.  You can expand, simplify, enable, disable, refactor any/all database calls you want on the PHP level, and the database will continue to log changes to preferred_names.  (we had discussed this at an internal meeting)

The only reason why there is a query comment is to show _who_ did the change.  We can't get that datapoint from the database because all database operations are conducted by the same DB user (submitty_dbuser).  And if we do lose this one piece in a PHP refactor, we aren't losing all functionality.  Preferred name changes will continue to be logged — we just won't know who did the change.  But we will have everything else.  Compare that to moving all logging to PHP — then we do risk losing all functionality on a refactor.

BTW, I did explain the query comment in code comments...  hopefully a developer who is paying attention won't refactor that out.  Regardless, the sysadmin tool portion validates the existence of the `AUTH` token (in the query comment).  I can go a step further and have it warn me when that token is missing, and follow up with a new issue to have that fixed.

**Comment by MasterOdin** [comment]
Yeah, let's make the script warn (but not fail) when it hits entries that lack the auth bit as I think it's vital here to have that information of who/what changed the field, and not just that it changed.

File: migration/migrator/data/submitty_db.sql [hunk 1]
```diff
@@ -51,6 +51,7 @@ CREATE TABLE courses (
 CREATE TABLE emails (
     id serial NOT NULL,
     user_id character varying NOT NULL,
+    recipient  varchar(255) NOT NULL,
```
Comment (MasterOdin) [inline migration/migrator/data/submitty_db.sql]: This probably shouldn't be re-added? (Removed in #4148)

**Comment by bmcutler** [comment]
Peter and I talked today,
he's going to try to get this automatically installed (both on vagrant & on production installations).
* clone the sysadmintools repo (like we clone rainbowgrades, lichen, etc)
* hopefully we can get the postgres logs created with read-only access by the submitty_daemon user (not world readable of course!)
* hook up the processing of the postgres logs to be done as a submitty_daemon nightly cron job
* write the preferred name logs to a new folder something like: /var/local/submitty/logs/preferred_names

(hopefully I got that mostly right)

File: migration/migrator/data/submitty_db.sql (hunk 1 above)
Comment (pbailie) [reply migration/migrator/data/submitty_db.sql]: I haven't actually touched the emails table.  Not sure how/why that is re-added.  But yeah — I'll take it back out.

File: .setup/install_system.sh [hunk 2]
```diff
@@ -762,6 +772,41 @@ if [[ ${VAGRANT} == 1 ]]; then
...
+function check_exit_code {
+# $1 = process return code
+# $2 = name of process
+    if [[ $1 -ne 0 ]]; then
+        echo -e "Error setting up $2.  Aborting."
+        exit 1
+    else
+        echo -e "Successfully set up $2."
+    fi
+}
```
Comment (MasterOdin) [inline .setup/install_system.sh]: this file uses `set -ev` at the top of it so any line that has a non-zero exit code is already going to cause it to crash and die, as well as print the line it failed on immediately before a big error message. If we really need this, do `set +e` around this chunk, but my thought here is that we don't and can remove this function and its usages.

File: .setup/install_system.sh [hunk 3]
```diff
@@ -762,6 +772,41 @@ if [[ ${VAGRANT} == 1 ]]; then
...
+# Copy preferred_name_logging.php to sbin
+rsync -qt ${SUBMITTY_REPOSITORY}/../SysadminTools/preferred_name_logging/preferred_name_logging.php ${SUBMITTY_INSTALL_DIR}/sbin
+check_exit_code $? "preferred_name_user.php"
+chown root:${DAEMON_GROUP} ${SUBMITTY_INSTALL_DIR}/sbin/preferred_name_logging.php
+chmod 0550 ${SUBMITTY_INSTALL_DIR}/sbin/preferred_name_logging.php
+
+# Adjust/overwrite Postgresql's configuration
+sed -i "s~^#*[ tab]*log_destination[ tab]*=[ tab]*'[a-z]\+'~log_destination = 'csvlog'~;
+        s~^#*[ tab]*logging_collector[ tab]*=[ tab]*[a-z01]\+~logging_collector = on~;
+        s~^#*[ tab]*log_directory[ tab]*=[ tab]*'[^][(){}<>|:;&#=!'?\*\~\$\"\` tab]\+'~log_directory = '/var/local/submitty/logs/psql'~;
```
[the remark this answers is not in the document]
Comment (pbailie) [reply .setup/install_system.sh]: I missed one.  Fixed.

File: .setup/install_system.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (pbailie) [reply .setup/install_system.sh]: OK — don't need this, then.  Removed in commit e3559ee.

File: .setup/install_system.sh [hunk 4]
```diff
@@ -762,6 +772,41 @@ if [[ ${VAGRANT} == 1 ]]; then
...
+        s~^#*[ tab]*log_filename[ tab]*=[ tab]*'[-a-zA-Z0-9_%\.]\+'~log_filename = 'postgresql-%Y-%m-%d_%H%M%S.log'~;
+        s~^#*[ tab]*log_file_mode[ tab]*=[ tab]*[0-9]\+~log_file_mode = 0640~;
+        s~^#*[ tab]*log_rotation_age[ tab]*=[ tab]*[a-z0-9]\+~log_rotation_age = 1d~;
+        s~^#*[ tab]*log_rotation_size[ tab]*=[ tab]*[a-zA-Z0-9]\+~log_rotation_size = 10MB~;
+        s~^#*[ tab]*log_min_messages[ tab]*=[ tab]*[a-z]\+~log_min_messages = log~;
+        s~^#*[ tab]*log_min_duration_statement[ tab]*=[ tab]*[-0-9]\+~log_min_duration_statement = 0~;
+        s~^#*[ tab]*log_line_prefix[ tab]*=[ tab]*'.\+'~log_line_prefix = '%t '~" /etc/postgresql/10/main/postgresql.conf
```
[the remark this answers is not in the document]
Comment (pbailie) [reply .setup/install_system.sh]: Done.

[the remark this answers is not in the document]
Comment (pbailie) [reply .setup/install_system.sh]: Turns out that this actually breaks the Vagrant dev VM, so I have moved the `sed` block (psql configuration) back to where it originally was.

The reason is that line 3 of the `sed` block sets 
```log_directory = '${SUBMITTY_DATA_DIR}/logs/psql'`.  This change request then has this configuration set _before_ the folder actually exists, and then issues a restart of postgresql.  This results in postgresql being shutdown due to an improper config.  The Vagrant VM eventually crashes during `# Create and setup database for non-workers` process.

**Comment by pbailie** [comment]
This PR also requires review and merging of Submitty/SysadminTools#19 and a new release point of the SysadminTools repository.

**Pull request merged**